lua/ccc/highlighter.lua

```diff
@@ -38,107 +37,143 @@ function Highlighter:init()
-    api.nvim_buf_clear_namespace(0, self.ns_id, 0, -1)
-    if not self.ft_filter[vim.bo.filetype] then
-        return
+---@param bufnr integer
+function Highlighter:start(bufnr)
+    for i, line in ipairs(api.nvim_buf_get_lines(bufnr, 0, -1, false)) do
+        self:update_line(bufnr, i - 1, line)
     end
-    local start_row = fn.line("w0") - 1
-    local end_row = fn.line("w$") - 1
-    for i, line in ipairs(api.nvim_buf_get_lines(0, start_row, end_row + 1, false)) do
-        local row = start_row + i - 1
-        local init = 1
-        while true do
-            local start, end_, RGB
-            for _, picker in ipairs(self.pickers) do
-                local s_, e_, rgb = picker.parse_color(line, init)
-                if s_ and (start == nil or s_ < start) then
-                    start = s_
-                    end_ = e_
-                    RGB = rgb
-                end
-            end
-            if start == nil then
-                break
+    if self.lsp then
+        self:update_lsp(bufnr)
+    end
+end
+
+---@param bufnr integer
+---@param first_line integer 0-index
+---@param last_line integer 0-index
+function Highlighter:update(bufnr, first_line, last_line)
+    vim.schedule(function()
+        api.nvim_buf_clear_namespace(bufnr, self.picker_ns_id, first_line, last_line)
+        for i, line in ipairs(api.nvim_buf_get_lines(bufnr, first_line, last_line, false)) do
+            local row = first_line + i - 1
+            self:update_line(bufnr, row, line)
+        end
+        if self.lsp then
+            self:update_lsp(bufnr)
+        end
+    end)
+end
+
+---@param bufnr integer
+---@param row integer 0-index
+---@param line string
+function Highlighter:update_line(bufnr, row, line)
+    local init = 1
+    while true do
+        local start, end_, RGB
+        for _, picker in ipairs(self.pickers) do
+            local s_, e_, rgb = picker.parse_color(line, init)
+            if s_ and (start == nil or s_ < start) then
+                start = s_
+                end_ = e_
+                RGB = rgb
             end
-            ---@cast RGB number[]
-            local hex = rgb2hex(RGB)
+        end
+        if start == nil then
+            break
+        end
+        ---@cast RGB number[]
+        local hex = rgb2hex(RGB)
+        local hl_name = "CccHighlighter" .. hex:sub(2)
+        if not self.is_defined[hex] then
+            local highlight = utils.create_highlight(hex)
+            api.nvim_set_hl(0, hl_name, highlight)
+            self.is_defined[hex] = true
+        end
+        api.nvim_buf_add_highlight(bufnr, self.picker_ns_id, hl_name, row, start - 1, end_)
+        init = end_ + 1
+    end
+end
+
+---@param bufnr integer
+function Highlighter:update_lsp(bufnr)
+    api.nvim_buf_clear_namespace(bufnr, self.lsp_ns_id, 0, -1)
+    local param = { textDocument = vim.lsp.util.make_text_document_params() }
+    vim.lsp.buf_request(bufnr, "textDocument/documentColor", param, function(err, colors)
```

Comment:
I don't recommend to use `vim.lsp.buf_request`.

If multiple lsp servers were attached to the target buffer, this method will callback multiple times.

I recommend to use `vim.lsp.get_active_clients()` and you should check servers capabilities.

Reply:
Is this correct what you are saying?